src/hotspot/share/nmt/indexedFreeListAllocator.hpp

```diff
@@ -95,103 +76,24 @@ class IndexedFreeListAllocator {
   }
 
   void free(I i) {
-    assert(!i.is_nil() || (i._idx > 0 && i._idx < _backing_storage.length()), "out of bounds free");
-    if (i.is_nil()) return;
-    BackingElement& be_freed = _backing_storage.at(i._idx);
+    assert(i != nil || (i > 0 && i < _backing_storage.length()), "out of bounds free");
```

Comment:
I think there are some errors here. This is probably broken. Which we would see if the gtests were running, but hotspot common tier1 tests seem broken.

Do we allow passing in nil? Then, i must be either nil or valid, not != nil or valid. If not, use an AND, not an OR.
i=0 is valid
Could you also please factor out OOB test for i?

Reply:
It's meant to express `P => Q` equiv. to `~P V Q`, but we're trying to say "is_not_nil(i) => is_valid(i)`, so it should be `i == nil` **not** `i != nil`. Yes, it's broken. Also yes, let's make OOB check a predicate function.